code/modules/client/preferences.dm

```diff
@@ -492,3 +500,226 @@ GLOBAL_LIST_EMPTY(preferences_datums)
 			default_randomization[preference_key] = RANDOM_ENABLED
 
 	return default_randomization
+
+/datum/preferences/proc/handle_client_importing(raw_data)
+	if(IsAdminAdvancedProcCall())
+		return
+
+	last_import_error_map = list()
+	if(!istext(raw_data) || !length(raw_data))
+		last_import_error_map[PREFERENCE_IMPORT_ERROR_INVALID_JSON] = TRUE
+		return
+
+	try
+		raw_data = json_decode(raw_data)
+	catch
+		last_import_error_map[PREFERENCE_IMPORT_ERROR_INVALID_JSON] = TRUE
+		return
+
+	var/raw_data_version = raw_data["version"]
+	var/imported_data_version = isnum(raw_data_version) ? raw_data_version : text2num(raw_data_version)
+	if(imported_data_version < PREFERENCES_VERSION_MINIMUM)
+		to_chat(parent, span_userdanger("Cannot import this savefile, it is too old!"))
+		last_import_error_map[PREFERENCE_IMPORT_ERROR_VERSION] += list(imported_data_version)
+		return
+
+	if(imported_data_version != PREFERENCES_VERSION_CURRENT)
+		if(tgui_alert(
+			parent,
+			"Importing a savefile which is older, this might cause issues.",
+			"Savefile Import",
+			list("Attempt Import", "Cancel"),
+		) != "Attempt Import")
+			return
+
+	var/static/regex/character_slot_regex = regex(@"^character[0-9]+$")
+	for(var/data_key in raw_data)
+		if(character_slot_regex.Find(data_key))
+			var/slot_id = text2num(replacetext(data_key, "character", ""))
+			if(slot_id > max_save_slots)
+				if(PREFERENCE_IMPORT_ERROR_NOT_ENOUGH_CHARACTER_SLOTS in last_import_error_map)
```

Comment:
@ZephyrTFA Wait, are we saving/loading ALL your characters into preferences? Seems like a major footgun if this can't be controlled...I might want to import one character over to another server and learn that all my characters are now replaced.

Is this trying to load into the same save slot as saved? If so that also seems silly--I wouldn't expect that at all and that would also clobber a character.

It should ask you what character(s) you want to replace, or if you want to use a new one. I don't want to clobber characters